src/ng/parse.js

```diff
@@ -622,15 +622,51 @@ function isStateless($filter, filterName) {
   return !fn.$stateful;
 }
 
-function findConstantAndWatchExpressions(ast, $filter) {
+// Detect nodes which could depend on non-shallow state of objects
+function isShallowState(stack) {
+  for (var i = 0; i < stack.length; i++) {
+    var node = stack[i];
+    switch (node.type) {
+      // Computed members might invoke a stateful toString()
+      case AST.MemberExpression:
+        if (node.computed) {
+          return false;
+        }
+        break;
+
+      // The + operator can invoke a stateful toString().
+      // Otherwise operators can not access any state within objects.
+      case AST.UnaryExpression:
+      case AST.BinaryExpression:
+        if (node.operator === '+') {
+          return false;
+        }
+        break;
+
+      // Functions / filters probably read state from within objects
+      case AST.CallExpression:
+        return false;
+    }
+  }
+
+  return true;
+}
+
+function findConstantAndWatchExpressions(ast, $filter, stack) {
```

Comment:
Yeah I think that's possible. I guess the idea of the `stack` was that it could also potentially be used for other purposes, but I'll try your suggestion.